clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -210,11 +204,27 @@ public RecordAppendResult append(TopicPartition tp,
     }
 
     /**
+     * If `RecordBatch.tryAppend` fails (i.e. the record batch is full), close its memory records to release temporary
+     * resources (like compression streams buffers).
+     */
+    private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, Callback callback, Deque<RecordBatch> deque) {
+        RecordBatch last = deque.peekLast();
+        if (last != null) {
+            FutureRecordMetadata future = last.tryAppend(timestamp, key, value, callback, time.milliseconds());
+            if (future == null)
+                last.records.close();
```

Comment:
`compressor.close()` will re-fill the wrapper message metadata by re-computing crc, etc; it does not result in correctness issues, but may incur unnecessary costs, since they are not just "no-op" when called again.